language/tools/read-write-set/types/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+mod access;
+
+pub use access::Access;
+
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::{TypeTag, ResourceKey},
+};
+use move_binary_format::normalized::Type;
+use std::{
+    collections::btree_map::BTreeMap,
+    fmt::{self, Formatter},
+};
+
+/// Offset of an access path: either a field, vector index, or global key
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Offset {
+    /// Index into contents of a struct by field offset
+    Field(usize),
+    /// Unknown index into a vector
+    VectorIndex,
+    /// A type index into global storage. Only follows a field or vector index of type address
+    Global(Type),
+}
+
+#[derive(Debug, Clone)]
+pub struct TrieNode {
+    /// Optional data associated with the parent in the trie
+    data: Option<Access>,
+    /// Child pointers labeled by offsets
+    children: BTreeMap<Offset, TrieNode>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Root {
+    Const(AccountAddress),
+    Formal(usize),
+}
+
+#[derive(Debug, Clone)]
+pub struct AccessPath {
+    pub root: Root,
+    pub offsets: Vec<Offset>,
+}
+
+#[derive(Debug, Clone)]
+pub struct ReadWriteSet(BTreeMap<Root, TrieNode>);
+
+impl AccessPath {
+    pub fn offset(&self) -> &[Offset] {
+        self.offsets.as_slice()
+    }
+    pub fn root(&self) -> &Root {
+        &self.root
+    }
+    pub fn add_offset(&mut self, offset: Offset) {
+        self.offsets.push(offset)
+    }
+    pub fn new_global_constant(addr: AccountAddress, ty: Type) -> Self {
+        Self {
+            root: Root::Const(addr),
+            offsets: vec![Offset::Global(ty)],
+        }
+    }
+    pub fn to_resource_key(&self) -> Option<ResourceKey> {
+        if self.offsets.len() != 1 {
```

Comment:
```suggestion
        if self.offsets.len() < 1 {
```

though I'm not sure what the intention of this function is. I would expect it to extract the resource key prefix from the given path; e.g. returning `None` for `0x1::M/f` instead of `Some(0x1::M)` seems surprising.

Reply:
I'm a bit concerned about the access path containing a seconadry index. Maybe I'll add another check here instead of simply asserting the path length to be 1.

Reply:
Check added!